pkg/storegateway/indexheader/reader_pool.go

```diff
@@ -149,7 +192,11 @@ func (p *ReaderPool) NewBinaryReader(ctx context.Context, logger log.Logger, bkt
 	}
 
 	if p.lazyReaderEnabled {
-		reader, err = NewLazyBinaryReader(ctx, readerFactory, logger, bkt, dir, id, p.metrics.lazyReader, p.onLazyReaderClosed)
+		lazyBinaryReader, lazyErr := NewLazyBinaryReader(ctx, readerFactory, logger, bkt, dir, id, p.metrics.lazyReader, p.onLazyReaderClosed)
+		if p.eagerLoadReaderEnabled {
```

Comment:
Design-wise, it feels a bit weird that we're passing the snapshot to the reader here - to me, the `ReaderPool` has the responsibility of managing what index-headers are loaded and persisting and loading the snapshot. 

Perhaps we should shift things around so the `ReaderPool` checks if the index-header should be eagerly loaded, and if it should be loaded, the `ReaderPool` instructs the `LazyBinaryReader` to load it? Then the only thing with knowledge of the snapshot and eager loading etc. is the `ReaderPool`.

Reply:
Updated as suggested. However I still need to pass the `usedAt` timestamp.

```
func (r *LazyBinaryReader) EagerLoadIndexHeadersSnapshot(usedAtSnapshot int64) 
```

Reply:
The usedAt argument might be not relevant anymore.

The above reply doesn't hold anymore after https://github.com/grafana/mimir/pull/5596#discussion_r1280437357 about setting usedAt to time.Now below. Hence the check need to be done in the caller and EagerLoad method will not receive any argument at all.